src/com/sap/piper/Utils.groovy

```diff
@@ -123,3 +123,20 @@ static String downloadSettingsFromUrl(script, String url, String targetFile = 's
     script.writeFile(file: targetFile, text: settings.getContent())
     return targetFile
 }
+
+/*
+ * Uses the Maven Help plugin to evaluate the given expression into the resolved values
+ * that maven sees at / generates at runtime. This way, the exact Maven coordinates and
+ * variables can be used.
+ */
+static String evaluateFromMavenPom(Script script, String pomFileName, String pomPathExpression) {
+
+    String resolvedExpression = script.mavenExecute(
```

Comment:
The longer I look at that line of code the more I think there is something wrong, depite it might work technically. Is it a good idea to use steps from the `vars` folder from the `src` folder?

We have the flow in the other direction (steps inside `vars` are calling utility functions in `src`). That are some kind of a dependencies. It is normally not a good idea to have backlinks into the other direction (some kind of cyclic dependencies).  

Maybe it would be better to have some maven utility interacting with maven and using this maven util from here (in this case there is no backlink to a step) and also from the `mavenExecute` step. In this case there is also no issue with having a return value from the`mavenExecute` step (see my other comment wrt return value) .

Reply:
I agree, but how should I implement that without re-implementing mavenExecute/dockerExecute ? 

We cannot expect that users have maven installed on their Jenkins/Jenkins-Node therefore we need to use a container with maven installed.

Reply:
It's java a prerequisite for running Jenkins?

Reply:
> It's java a prerequisite for running Jenkins?

[Yes](https://jenkins.io/doc/administration/requirements/java/).

By the way: I do not fully understand how the questions fits into the discussion here.

Reply:
> It's java a prerequisite for running Jenkins?

I am writing about `maven` not `java` and `mvn` is not available in the cx-jenkins-master container.

Reply:
And it is a hard requirement on Cloud SDK side that we don't assume you're installing mvn and other tools in the Jenkins, as our whole project is based on the idea of using containers for various reasons.

Reply:
> we don't assume you're installing mvn

No, no, no local mvn installation, that's fully agreed ...

We have already some cases where we go back from `src` to `vars`. In general I still think we should avoid that , but we have already this cases. Details can be checked via

```
for step in `ls -la vars |sed 's/^.* //g' |sed 's/.groovy$//g' |sed -e '1,3d'  `; do echo ===== $step ====== ; grep -r  $step src ; done
```

Basically we have:

- commonPipelineEnvironment: src/com/sap/piper/ConfigurationHelper.groovy
- dockerExecute:  src/com/sap/piper/cm/ChangeManagement.groovy
- mavenExecute: src/com/sap/piper/versioning/MavenArtifactVersioning.groovy

Not sure if we should do a bigger re-work now in order to avoid the callbacks from `src` to `vars`. In general I think we should avoid it, but maybe it is justified to go forward now with the current approach.

Reply:
> > It's java a prerequisite for running Jenkins?
> 
> I am writing about `maven` not `java` and `mvn` is not available in the cx-jenkins-master container.

Sorry, mixed it up.